### DISCUSSION TITLE
消息通知模块设计

### DISCUSSION THREAD

**Discussion by fanhousanbu** [go-atomci/atomci#51] [General]
### 模块目标
> 流水线的关键节点（启动或完成）执行结果及时通知关注者
### 消息组件
> 考虑集成开源项目
### 模块设计
按产研团队配置分阶段实施
#### 阶段一：
1. **适用场景**：团队产研人员基本固定、且**人数较少**（估且认为10人以内）
2. **解决方案**：系统中配置通知目标；在项目流程中配置通知节点；流水线执行过程中自动触发通知
![image](https://user-images.githubusercontent.com/6702542/149279700-95111184-a7b4-47ff-9d32-419964a84793.png)

#### 阶段二：
1. **适用场景**：团队产研人员不固定，且**人数较少**的团队（这应该是主流用户群体）
2. **解决方案**：在流水线创建时配置通知目标
![image](https://user-images.githubusercontent.com/6702542/149279709-ea1647b5-5acf-4dd9-99e3-3ea7bed66ade.png)

#### 阶段三、四：
1. **适用场景**：团队产研人数较多
2. **解决方案**：未知，但阶段三的前置条件应该是AtomCI具备成员管理能力

### Refer To
1. [消息通知模块设计](https://www.yuque.com/docs/share/8b428d01-8501-44ac-bd5e-cc2a45be690c?)

**Comment by colynn** [comment]
@fanhousanbu  感谢，很好的建议方案，想确认下： 阶段二中『创建流水线』可以理解为 『项目流程』的一次实例化，所以如方案说描述的 创建流水线 添加通知对象，那么消息通知与 创建的流水线实例，项目流程 他们三者之间是什么关系？

**Reply by fanhousanbu** [reply]
目前我对通知在这套系统中的理解还仅限于一个通知工具，所以，这三者的关注用伪代码来表达，差不多是：
```
var pipeline = new Flow(INotify noti);

class Flow{
   public Flow(INotify noti){
        this._noti = noti;
  }

  public void DeploySuccess(){
       // biz logic
      // noti
     _noti.Send(...);
 }
}
```

之前讨论过，参考云效的通知作为节点的设计，目前的理解不够深入，所以暂时表达不出更多的含义了

**Comment by colynn** [comment]
From 叶飞： 通知 可以是 `webhook`能力上的子集

**Reply by fanhousanbu** [reply]
> From 叶飞： 通知 可以是 `webhook`能力上的子集

不应该这么定义吧？虽然底层都用webhook ，但他们的关注点不同，套用DDD的说法，同一个名词在两个领域可能就完全是两回事
